doc/format-run.md

```diff
@@ -212,16 +212,26 @@ read first. This helps with pre-allocating the memory needed for the
 in-memory representation of the other main components, and knowing how much
 data to read from disk for each component.
 
+The clash map is expected to be very small, so its file format is mainly
+designed to be simple, not compact. After a 64bit number of elements, each pair
+of key and page number is serialised in the following order:
+* the page number (as a 32bit value)
+* the length of the key in bytes (as a 32bit value)
+* the key
+* padding to 64bit alignment
+
 Size granularity, alignment and any trailing padding of the components:
 1. 32bit size granularity, 32bit alignment
 2. 32bit size granularity, 32bit alignment, trailing padding to 64bit alignment
 3. 64bit size granularity, 64bit alignment
 4. 64bit size granularity, 64bit alignment
-5. 8bit size granularity, 8bit alignment, trailing padding to 64bit alignment
+5. 32bit size granularity, 32bit alignment, trailing padding to 64bit alignment
```

Comment:
Not sure if that description fits the format. Internally, there are 32 bit numbers and even arbitrary-byte sized keys, but each individual element (pair) of the clash map has padding to 64 bit.

Reply:
What is important here is the alignment that the component ensures for the remaining data. I originally wrote it as 8bit and 8bit because that's the weakest promise we can make about an at-the-time unspecified format.

There's two sensible strategies for the format of the clash map:
1. treat it as a separate byte string that gets deserialised into some other structure: thus the internal alignment doesn't matter since one uses a "proper" parser to read it
2. treat it like the other arrays, that are accessed in-place when the data is copied into memory, without any separate deserialisation step

I had originally been assuming strategy 1, and thus thinking of the format spec as a bytestring with no particular alignment, except padding at the end.

Reply:
I first had a prototype using the `Binary` package, but it required making a few orphan instances. It didn't feel like it made things simpler, although I did have to manually add bounds checking now. :shrug: